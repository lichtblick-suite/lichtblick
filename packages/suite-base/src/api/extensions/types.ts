// SPDX-FileCopyrightText: Copyright (C) 2023-2025 Bayerische Motoren Werke Aktiengesellschaft (BMW AG)<lichtblick@bmwgroup.com>
// SPDX-License-Identifier: MPL-2.0

import { StoredExtension } from "@lichtblick/suite-base/services/IExtensionStorage";
import { ExtensionInfo } from "@lichtblick/suite-base/types/Extensions";

export interface IExtensionAPI {
  createOrUpdate(extension: ExtensionInfoSlug, file: File): Promise<StoredExtension>;
  get(id: string): Promise<StoredExtension | undefined>;
  loadContent(fileId: string): Promise<Uint8Array | undefined>;
  list(): Promise<ExtensionInfo[]>;
  remove(id: string): Promise<boolean>;
  readonly slug: string;
}

export type ExtensionInfoSlug = Pick<StoredExtension, "info" | "slug">;
