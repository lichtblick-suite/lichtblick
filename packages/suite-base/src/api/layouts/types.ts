// SPDX-FileCopyrightText: Copyright (C) 2023-2025 Bayerische Motoren Werke Aktiengesellschaft (BMW AG)<lichtblick@bmwgroup.com>
// SPDX-License-Identifier: MPL-2.0

import { LayoutData, LayoutID } from "@lichtblick/suite-base/context/CurrentLayoutContext";
import {
  ISO8601Timestamp,
  Layout,
  LayoutPermission,
} from "@lichtblick/suite-base/services/ILayoutStorage";
import { RemoteLayout } from "@lichtblick/suite-base/services/IRemoteLayoutStorage";

export type SaveNewLayout = Pick<Layout, "name" | "data" | "permission" | "externalId"> & {
  id: LayoutID | undefined;
  savedAt: ISO8601Timestamp;
};

export type UpdateLayoutRequest = {
  id: LayoutID;
  externalId: string;
  name?: string;
  data?: LayoutData;
  permission?: LayoutPermission;
  savedAt: ISO8601Timestamp;
};

export type UpdateLayoutResponse =
  | { status: "success"; newLayout: RemoteLayout }
  | { status: "conflict" };

export type RemoteLayoutResponse = {
  data: RemoteLayoutResponseData[];
};

export type RemoteLayoutResponseData = {
  id: string;
  data: LayoutData;
  name: string;
  layoutId: LayoutID;
  namespace: string;
  permission: LayoutPermission;
  from: string;
  createdBy: string;
  updatedBy: string;
};
