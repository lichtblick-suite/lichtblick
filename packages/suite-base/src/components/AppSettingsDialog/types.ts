// SPDX-FileCopyrightText: Copyright (C) 2023-2025 Bayerische Motoren Werke Aktiengesellschaft (BMW AG)<lichtblick@bmwgroup.com>
// SPDX-License-Identifier: MPL-2.0

/**
 * @fileoverview AppSettingsDialog - 型定義
 *
 * 【概要】
 * AppSettingsDialogコンポーネントで使用される型定義を提供します。
 * 設定ダイアログのタブ管理と、Aboutセクションの構造を定義しています。
 *
 * 【主な型】
 * - `AppSettingsSectionKey` - Aboutセクションの項目キー
 * - `AppSettingsTab` - 設定ダイアログのタブ識別子
 *
 * 【使用箇所】
 * - AppSettingsDialog.tsx - メインコンポーネント
 * - constants.ts - 定数定義
 * - SettingsMenu.tsx - 設定メニュー
 * - WorkspaceContext.ts - ワークスペース状態管理
 */

/**
 * AppSettingsセクションキー
 *
 * Aboutタブ内で表示される情報セクションの識別子。
 * 各セクションは独立したカテゴリとして扱われ、
 * 対応するリンクや情報を持ちます。
 *
 * @example
 * ```typescript
 * const section: AppSettingsSectionKey = "documentation";
 * ```
 */
export type AppSettingsSectionKey = "documentation" | "legal";

/**
 * AppSettingsタブ
 *
 * 設定ダイアログで表示可能なタブの識別子。
 * 各タブは独立した設定カテゴリを表し、
 * 対応する設定項目のセクションを持ちます。
 *
 * タブの構成：
 * - `general` - 一般設定（テーマ、言語、タイムゾーンなど）
 * - `privacy` - プライバシー設定（テレメトリ、クラッシュレポートなど）
 * - `extensions` - 拡張機能の管理
 * - `experimental-features` - 実験的機能の有効/無効
 * - `about` - アプリケーション情報とリンク
 *
 * @example
 * ```typescript
 * const activeTab: AppSettingsTab = "general";
 *
 * // 特定のタブで設定ダイアログを開く
 * <AppSettingsDialog activeTab="extensions" />
 * ```
 */
export type AppSettingsTab =
  | "general"
  | "privacy"
  | "extensions"
  | "experimental-features"
  | "about";
