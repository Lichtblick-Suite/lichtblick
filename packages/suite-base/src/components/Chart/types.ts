// This Source Code Form is subject to the terms of the Mozilla Public
// License, v2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/

import { ScatterDataPoint, ChartData as ChartJsChartData } from "chart.js";

// Chartjs typings use _null_ to indicate _gaps_ in the dataset
// eslint-disable-next-line no-restricted-syntax
const ChartNull = null;
type Datum = ScatterDataPoint & {
  // chart.js supported properties to show a label above the datapoint
  // used by the state transition panel to show a label above the transition datum
  label?: string;
  labelColor?: string;

  // Our additional properties
  // value is the original value (rather than the plot x/y value) for the datum (used by state transitions)
  value?: string | number | bigint | boolean;
  // Constant name for the datum (used by state transitions)
  constantName?: string | undefined;

  // Contains all of the distinct states present in this line segment. Only for
  // state transition data.
  states?: string[];
};
export type ObjectData = (Datum | typeof ChartNull)[];
export type ChartData = ChartJsChartData<"scatter", ObjectData>;

export type TypedData = {
  x: Float32Array;
  y: Float32Array;
  value: (string | number | bigint | boolean | undefined)[];
  constantName?: string[];
};
export type TypedChartData = ChartJsChartData<"scatter", TypedData[]>;

export type RpcScale = {
  // min scale value
  min: number;
  // max scale value
  max: number;
  // pixel coordinate within the component that corresponds to min
  pixelMin: number;
  // pixel coordinate within the component that corresponds to max
  pixelMax: number;
};

export type RpcScales = {
  x?: RpcScale;
  y?: RpcScale;
};

export type RpcElement = {
  data?: Datum;
  datasetIndex: number;
  index: number;
  view: {
    x: number;
    y: number;
  };
};
