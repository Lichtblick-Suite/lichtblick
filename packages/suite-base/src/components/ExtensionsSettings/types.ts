// SPDX-FileCopyrightText: Copyright (C) 2023-2024 Bayerische Motoren Werke Aktiengesellschaft (BMW AG)<lichtblick@bmwgroup.com>
// SPDX-License-Identifier: MPL-2.0

import { AsyncState } from "react-use/lib/useAsyncFn";

import { Immutable } from "@lichtblick/suite";
import { ExtensionMarketplaceDetail } from "@lichtblick/suite-base/context/ExtensionMarketplaceContext";

export type InstalledExtension = {
  id: string;
  installed: boolean;
  name: string;
  displayName: string;
  description: string;
  publisher: string;
  homepage?: string;
  license?: string;
  version: string;
  keywords?: string[];
  namespace: string;
  qualifiedName: string;
};

export type FocusedExtension = {
  installed: boolean;
  entry: Immutable<ExtensionMarketplaceDetail>;
};

export type EntryGroupedData = {
  namespace: string;
  entries: Immutable<ExtensionMarketplaceDetail>[];
};

export type UseExtensionSettingsHook = {
  setUndebouncedFilterText: (newFilterText: string) => void;
  marketplaceEntries: AsyncState<ExtensionMarketplaceDetail[]>;
  refreshMarketplaceEntries: () => Promise<ExtensionMarketplaceDetail[]>;
  undebouncedFilterText: string;
  namespacedData: EntryGroupedData[];
  groupedMarketplaceData: EntryGroupedData[];
  debouncedFilterText: string;
};
