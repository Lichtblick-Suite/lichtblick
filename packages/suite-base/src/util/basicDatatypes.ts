// This Source Code Form is subject to the terms of the Mozilla Public
// License, v2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/

import { RosDatatypes } from "@lichtblick/suite-base/types/RosDatatypes";

import { ros1, ros2galactic } from "@foxglove/rosmsg-msgs-common";
import { foxgloveMessageSchemas, generateRosMsgDefinition } from "@foxglove/schemas/internal";

/**
 * basicDatatypes is a map containing definitions for ROS common datatypes and foxglove datatypes
 * from the following packages:
 *
 * - @foxglove/rosmsgs-msg-common
 * - @foxglove/schemas
 */
export const basicDatatypes: RosDatatypes = new Map();

// Add ROS2 message schemas
for (const [name, def] of Object.entries(ros2galactic)) {
  basicDatatypes.set(name, def);
}

// Add ROS1 message schemas
for (const [name, def] of Object.entries(ros1)) {
  basicDatatypes.set(name, def);
}

// Add foxglove message schemas
for (const schema of Object.values(foxgloveMessageSchemas)) {
  const definition = generateRosMsgDefinition(schema, { rosVersion: 1 });
  basicDatatypes.set(definition.rosMsgInterfaceName, {
    name: definition.rosMsgInterfaceName,
    definitions: definition.fields,
  });
  basicDatatypes.set(`foxglove.${schema.name}`, {
    name: `foxglove.${schema.name}`,
    definitions: definition.fields,
  });
}

// Add the legacy foxglove_msgs/ImageMarkerArray message definition
basicDatatypes.set("foxglove_msgs/ImageMarkerArray", {
  definitions: [
    { name: "markers", type: "visualization_msgs/ImageMarker", isComplex: true, isArray: true },
  ],
});
