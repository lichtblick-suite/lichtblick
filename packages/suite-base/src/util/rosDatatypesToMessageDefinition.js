// SPDX-FileCopyrightText: Copyright (C) 2023-2025 Bayerische Motoren Werke Aktiengesellschaft (BMW AG)<lichtblick@bmwgroup.com>
// SPDX-License-Identifier: MPL-2.0
// For one datatype in the datatypes, find the MessageDefinition that we can use
// to either write or parse it. `datatypes` should contain the root datatype and
// all complex sub-datatypes.
export default function rosDatatypesToMessageDefinition(datatypes, rootDatatypeName) {
    const result = [];
    const seenDatatypeNames = new Set([rootDatatypeName]);
    // It doesn't matter if we use a stack or queue here, but we use a stack.
    const datatypeNameStack = [rootDatatypeName];
    while (datatypeNameStack.length > 0) {
        const currentDatatypeName = datatypeNameStack.pop();
        if (currentDatatypeName == undefined) {
            throw new Error(`Invariant violation - Array.pop() when length > 0`);
        }
        const currentDatatype = datatypes.get(currentDatatypeName);
        if (!currentDatatype) {
            throw new Error(`While searching datatypes for "${rootDatatypeName}", could not find datatype "${currentDatatypeName}"`);
        }
        // The root datatype has no name field.
        const msgDefinition = currentDatatypeName === rootDatatypeName
            ? { definitions: currentDatatype.definitions }
            : { name: currentDatatypeName, definitions: currentDatatype.definitions };
        result.push(msgDefinition);
        for (const field of currentDatatype.definitions) {
            // Only search subfields if we haven't already seen it and it is "complex", IE it has its own fields and should
            // be contained in `datatypes`.
            if (field.isComplex === true && !seenDatatypeNames.has(field.type)) {
                datatypeNameStack.push(field.type);
                seenDatatypeNames.add(field.type);
            }
        }
    }
    return result;
}
