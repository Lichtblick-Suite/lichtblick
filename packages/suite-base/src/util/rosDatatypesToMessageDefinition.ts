// This Source Code Form is subject to the terms of the Mozilla Public
// License, v2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/
//
// This file incorporates work covered by the following copyright and
// permission notice:
//
//   Copyright 2018-2021 Cruise LLC
//
//   This source code is licensed under the Apache License, Version 2.0,
//   found at http://www.apache.org/licenses/LICENSE-2.0
//   You may not use this file except in compliance with the License.

import { RosDatatypes } from "@lichtblick/suite-base/types/RosDatatypes";

import { MessageDefinition } from "@foxglove/message-definition";

// For one datatype in the datatypes, find the MessageDefinition that we can use
// to either write or parse it. `datatypes` should contain the root datatype and
// all complex sub-datatypes.
export default function rosDatatypesToMessageDefinition(
  datatypes: RosDatatypes,
  rootDatatypeName: string,
): MessageDefinition[] {
  const result = [];
  const seenDatatypeNames = new Set([rootDatatypeName]);
  // It doesn't matter if we use a stack or queue here, but we use a stack.
  const datatypeNameStack = [rootDatatypeName];

  while (datatypeNameStack.length > 0) {
    const currentDatatypeName = datatypeNameStack.pop();
    if (currentDatatypeName == undefined) {
      throw new Error(`Invariant violation - Array.pop() when length > 0`);
    }
    const currentDatatype = datatypes.get(currentDatatypeName);
    if (!currentDatatype) {
      throw new Error(
        `While searching datatypes for "${rootDatatypeName}", could not find datatype "${currentDatatypeName}"`,
      );
    }
    // The root datatype has no name field.
    const msgDefinition: MessageDefinition =
      currentDatatypeName === rootDatatypeName
        ? { definitions: currentDatatype.definitions }
        : { name: currentDatatypeName, definitions: currentDatatype.definitions };
    result.push(msgDefinition);
    for (const field of currentDatatype.definitions) {
      // Only search subfields if we haven't already seen it and it is "complex", IE it has its own fields and should
      // be contained in `datatypes`.
      if (field.isComplex === true && !seenDatatypeNames.has(field.type)) {
        datatypeNameStack.push(field.type);
        seenDatatypeNames.add(field.type);
      }
    }
  }

  return result;
}
