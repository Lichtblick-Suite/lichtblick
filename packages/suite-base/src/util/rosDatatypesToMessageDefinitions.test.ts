// This Source Code Form is subject to the terms of the Mozilla Public
// License, v2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/
//
// This file incorporates work covered by the following copyright and
// permission notice:
//
//   Copyright 2018-2021 Cruise LLC
//
//   This source code is licensed under the Apache License, Version 2.0,
//   found at http://www.apache.org/licenses/LICENSE-2.0
//   You may not use this file except in compliance with the License.

import { RosDatatypes } from "@lichtblick/suite-base/types/RosDatatypes";
import * as _ from "lodash-es";

import { basicDatatypes } from "./basicDatatypes";
import rosDatatypesToMessageDefinition from "./rosDatatypesToMessageDefinition";

describe("rosDatatypesToMessageDefinition", () => {
  it(`Includes all of the definitions for "visualization_msgs/StudioMarkerArray"`, () => {
    expect(
      rosDatatypesToMessageDefinition(basicDatatypes, "visualization_msgs/MarkerArray"),
    ).toMatchSnapshot();
  });

  it("produces a correct message definition", () => {
    const definitions = rosDatatypesToMessageDefinition(
      basicDatatypes,
      "visualization_msgs/MarkerArray",
    );
    // Should have 1 definition without a name, the root datatype.
    expect(definitions.filter(({ name }) => name == undefined).length).toEqual(1);
    // Should not duplicate definitions.
    expect(_.uniqBy(definitions, "name").length).toEqual(definitions.length);
  });

  it("Errors if it can't find the definition", () => {
    const datatypes: RosDatatypes = new Map(
      Object.entries({
        ["visualization_msgs/MarkerArray"]: {
          definitions: [
            {
              isArray: true,
              isComplex: true,
              arrayLength: undefined,
              name: "markers",
              type: "visualization_msgs/Marker",
            },
            {
              isArray: false,
              isComplex: true,
              name: "header",
              type: "std_msgs/CustomHeader",
            },
          ],
        },
      }),
    );
    expect(() =>
      rosDatatypesToMessageDefinition(datatypes, "visualization_msgs/MarkerArray"),
    ).toThrow(
      'While searching datatypes for "visualization_msgs/MarkerArray", could not find datatype "std_msgs/CustomHeader"',
    );
  });
});
