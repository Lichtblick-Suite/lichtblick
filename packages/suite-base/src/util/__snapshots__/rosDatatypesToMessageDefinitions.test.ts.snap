// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rosDatatypesToMessageDefinition Includes all of the definitions for "visualization_msgs/StudioMarkerArray" 1`] = `
[
  {
    "definitions": [
      {
        "isArray": true,
        "isComplex": true,
        "name": "markers",
        "type": "visualization_msgs/Marker",
      },
    ],
  },
  {
    "definitions": [
      {
        "isConstant": true,
        "name": "ARROW",
        "type": "uint8",
        "value": 0,
        "valueText": "0",
      },
      {
        "isConstant": true,
        "name": "CUBE",
        "type": "uint8",
        "value": 1,
        "valueText": "1",
      },
      {
        "isConstant": true,
        "name": "SPHERE",
        "type": "uint8",
        "value": 2,
        "valueText": "2",
      },
      {
        "isConstant": true,
        "name": "CYLINDER",
        "type": "uint8",
        "value": 3,
        "valueText": "3",
      },
      {
        "isConstant": true,
        "name": "LINE_STRIP",
        "type": "uint8",
        "value": 4,
        "valueText": "4",
      },
      {
        "isConstant": true,
        "name": "LINE_LIST",
        "type": "uint8",
        "value": 5,
        "valueText": "5",
      },
      {
        "isConstant": true,
        "name": "CUBE_LIST",
        "type": "uint8",
        "value": 6,
        "valueText": "6",
      },
      {
        "isConstant": true,
        "name": "SPHERE_LIST",
        "type": "uint8",
        "value": 7,
        "valueText": "7",
      },
      {
        "isConstant": true,
        "name": "POINTS",
        "type": "uint8",
        "value": 8,
        "valueText": "8",
      },
      {
        "isConstant": true,
        "name": "TEXT_VIEW_FACING",
        "type": "uint8",
        "value": 9,
        "valueText": "9",
      },
      {
        "isConstant": true,
        "name": "MESH_RESOURCE",
        "type": "uint8",
        "value": 10,
        "valueText": "10",
      },
      {
        "isConstant": true,
        "name": "TRIANGLE_LIST",
        "type": "uint8",
        "value": 11,
        "valueText": "11",
      },
      {
        "isConstant": true,
        "name": "ADD",
        "type": "uint8",
        "value": 0,
        "valueText": "0",
      },
      {
        "isConstant": true,
        "name": "MODIFY",
        "type": "uint8",
        "value": 0,
        "valueText": "0",
      },
      {
        "isConstant": true,
        "name": "DELETE",
        "type": "uint8",
        "value": 2,
        "valueText": "2",
      },
      {
        "isConstant": true,
        "name": "DELETEALL",
        "type": "uint8",
        "value": 3,
        "valueText": "3",
      },
      {
        "isArray": false,
        "isComplex": true,
        "name": "header",
        "type": "std_msgs/Header",
      },
      {
        "isArray": false,
        "isComplex": false,
        "name": "ns",
        "type": "string",
      },
      {
        "isArray": false,
        "isComplex": false,
        "name": "id",
        "type": "int32",
      },
      {
        "isArray": false,
        "isComplex": false,
        "name": "type",
        "type": "int32",
      },
      {
        "isArray": false,
        "isComplex": false,
        "name": "action",
        "type": "int32",
      },
      {
        "isArray": false,
        "isComplex": true,
        "name": "pose",
        "type": "geometry_msgs/Pose",
      },
      {
        "isArray": false,
        "isComplex": true,
        "name": "scale",
        "type": "geometry_msgs/Vector3",
      },
      {
        "isArray": false,
        "isComplex": true,
        "name": "color",
        "type": "std_msgs/ColorRGBA",
      },
      {
        "isArray": false,
        "isComplex": false,
        "name": "lifetime",
        "type": "duration",
      },
      {
        "isArray": false,
        "isComplex": false,
        "name": "frame_locked",
        "type": "bool",
      },
      {
        "isArray": true,
        "isComplex": true,
        "name": "points",
        "type": "geometry_msgs/Point",
      },
      {
        "isArray": true,
        "isComplex": true,
        "name": "colors",
        "type": "std_msgs/ColorRGBA",
      },
      {
        "isArray": false,
        "isComplex": false,
        "name": "text",
        "type": "string",
      },
      {
        "isArray": false,
        "isComplex": false,
        "name": "mesh_resource",
        "type": "string",
      },
      {
        "isArray": false,
        "isComplex": false,
        "name": "mesh_use_embedded_materials",
        "type": "bool",
      },
    ],
    "name": "visualization_msgs/Marker",
  },
  {
    "definitions": [
      {
        "isArray": false,
        "isComplex": false,
        "name": "x",
        "type": "float64",
      },
      {
        "isArray": false,
        "isComplex": false,
        "name": "y",
        "type": "float64",
      },
      {
        "isArray": false,
        "isComplex": false,
        "name": "z",
        "type": "float64",
      },
    ],
    "name": "geometry_msgs/Point",
  },
  {
    "definitions": [
      {
        "isArray": false,
        "isComplex": false,
        "name": "r",
        "type": "float32",
      },
      {
        "isArray": false,
        "isComplex": false,
        "name": "g",
        "type": "float32",
      },
      {
        "isArray": false,
        "isComplex": false,
        "name": "b",
        "type": "float32",
      },
      {
        "isArray": false,
        "isComplex": false,
        "name": "a",
        "type": "float32",
      },
    ],
    "name": "std_msgs/ColorRGBA",
  },
  {
    "definitions": [
      {
        "isArray": false,
        "isComplex": false,
        "name": "x",
        "type": "float64",
      },
      {
        "isArray": false,
        "isComplex": false,
        "name": "y",
        "type": "float64",
      },
      {
        "isArray": false,
        "isComplex": false,
        "name": "z",
        "type": "float64",
      },
    ],
    "name": "geometry_msgs/Vector3",
  },
  {
    "definitions": [
      {
        "isArray": false,
        "isComplex": true,
        "name": "position",
        "type": "geometry_msgs/Point",
      },
      {
        "isArray": false,
        "isComplex": true,
        "name": "orientation",
        "type": "geometry_msgs/Quaternion",
      },
    ],
    "name": "geometry_msgs/Pose",
  },
  {
    "definitions": [
      {
        "isArray": false,
        "isComplex": false,
        "name": "x",
        "type": "float64",
      },
      {
        "isArray": false,
        "isComplex": false,
        "name": "y",
        "type": "float64",
      },
      {
        "isArray": false,
        "isComplex": false,
        "name": "z",
        "type": "float64",
      },
      {
        "isArray": false,
        "isComplex": false,
        "name": "w",
        "type": "float64",
      },
    ],
    "name": "geometry_msgs/Quaternion",
  },
  {
    "definitions": [
      {
        "isArray": false,
        "isComplex": false,
        "name": "seq",
        "type": "uint32",
      },
      {
        "isArray": false,
        "isComplex": false,
        "name": "stamp",
        "type": "time",
      },
      {
        "isArray": false,
        "isComplex": false,
        "name": "frame_id",
        "type": "string",
      },
    ],
    "name": "std_msgs/Header",
  },
]
`;
